### PR TITLE
Harden package version detection from `git describe`

### PR DESCRIPTION
**Pull request by kinke**
Resolves issue #2262.

### CODE REVIEW

File: source/dub/internal/git.d [hunk 1]
```diff
@@ -102,6 +91,44 @@ private string determineVersionWithGitTool(NativePath path)
...
+	// 1 commit after v1.0.0
+	assert(determineVersionFromGitDescribe("v1.0.0-1-deadbeef") == "1.0.0-commit.1.deadbeef");
+	// tag v1.0.0+2.0.0
+	assert(determineVersionFromGitDescribe("v1.0.0+2.0.0-0-deadbeef") == "1.0.0+2.0.0");
+	// 12 commits after tag v1.0.0+2.0.0
+	assert(determineVersionFromGitDescribe("v1.0.0+2.0.0-12-deadbeef") == "1.0.0-commit.12.deadbeef+2.0.0");
+	// tag v1.0.0-beta.1
+	assert(determineVersionFromGitDescribe("v1.0.0-beta.1-0-deadbeef") == "1.0.0-beta.1");
+	// 2 commits after tag v1.0.0-beta.1
+	assert(determineVersionFromGitDescribe("v1.0.0-beta.1-2-deadbeef") == "1.0.0-beta.1-commit.2.deadbeef");
```
Comment (kinke) [inline source/dub/internal/git.d:125]: @Geod24: Could you please have a look at these and confirm that's all semver-compatible?

**Review by thewilsonator — APPROVED** [review]

File: source/dub/internal/git.d [hunk 2]
```diff
@@ -102,6 +91,44 @@ private string determineVersionWithGitTool(NativePath path)
...
+		auto r = format("%s-commit.%s.%s", tag[1 .. (i < 0) ? $ : i], num, commit);
+		if (i > 0) r ~= tag[i .. $];
+		return r;
+	}
+	return null;
+}
+
+unittest {
+	// tag v1.0.0
+	assert(determineVersionFromGitDescribe("v1.0.0-0-deadbeef") == "1.0.0");
```
Comment (Geod24) [inline source/dub/internal/git.d:116]: Note that the commit hash will always start with `g`. Don't know if you wish to rely on that or not.

File: source/dub/internal/git.d [hunk 3]
```diff
@@ -102,6 +91,49 @@ private string determineVersionWithGitTool(NativePath path)
...
+	// 1 commit after v1.0.0
+	assert(determineVersionFromGitDescribe("v1.0.0-1-deadbeef") == "1.0.0-commit.1.deadbeef");
+	// tag v1.0.0+2.0.0
+	assert(determineVersionFromGitDescribe("v1.0.0+2.0.0-0-deadbeef") == "1.0.0+2.0.0");
+	// 12 commits after tag v1.0.0+2.0.0
+	assert(determineVersionFromGitDescribe("v1.0.0+2.0.0-12-deadbeef") == "1.0.0-commit.12.deadbeef+2.0.0");
```
Comment (Geod24) [inline source/dub/internal/git.d:122]: What happens with plain `v1.0.0+2.0.0` ?

Comment (Geod24) [reply source/dub/internal/git.d:122]: Likewise, `v1.0.0-alpha+alpn`

**Review by Geod24 — COMMENTED** [review]
It's better, but I don't think it's the right solution.
[Using a hyphen denotes a pre-release version](https://semver.org/#spec-item-9).
Obviously, those versions are not pre-release. They are, if anything, post-release.

The only benefit we derive from using a pre-release approach is that version comparison will favor the final version over those. But again, it can lead to surprising results.

So something such as `v1.1.1-26-g6caf458` would be more suited as one of:
- `v1.1.2-26-g6caf458`
- `v1.2.0-26-g6caf458`
- `v2.0.0-26-g6caf458`

There are arguments for all three choices. I'd stay clear of the last one, as it's too purist to be of use, and would consider the second one over the first one, because it's the most likely scenario (new features).

The last approach would also to just say "dang it" and use the commit hash and expect no version match to work.

File: source/dub/internal/git.d (hunk 3 above)
Comment (kinke) [reply source/dub/internal/git.d:122]: That's no valid `git describe --long --tags` output.

Comment (Geod24) [reply source/dub/internal/git.d:122]: If it's run on a tag, you get whatever the tag is, or am I missing something ?

**Comment by kinke** [comment]
> The only benefit we derive from using a pre-release approach is that version comparison will favor the final version over those.

The (important) thing that is fixed by this is that a `1.0.0` selection isn't satisfied by some existing `1.0.0+…` git+https cache entry anymore. AFAIU, it also shouldn't match a `~>1.0` or `>=1.0.0` anymore. But a `1.0.0-commit.N.<hash>` would probably still match some `>=1.0.0-beta.1`, but I guess that's not very likely to cause trouble.

> The last approach would also to just say "dang it" and use the commit hash and expect no version match to work.

I haven't tested it, but was expecting loads of `selection <bla> doesn't match requirement <blub>` warnings in that case.

File: source/dub/internal/git.d (hunk 3 above)
Comment (kinke) [reply source/dub/internal/git.d:122]: Nope, I've tested it - you get `<tag>-0-<hash>` in that case as well. Hence the `0` special case in the original code as well.

Comment (Geod24) [reply source/dub/internal/git.d:122]: Perfect, thanks

**Review by Geod24 — APPROVED** [review]
Makes sense, LGTM

**Comment by s-ludwig** [comment]
This is actually introduces a pretty bad regression. Mapping the following commits to the latest version is very much on purpose and an essential part of working with multiple dependent packages side by side. Also, `1.0.0-commit.1fa6345` is not a valid SemVer version, because the last segment would be interpreted as a number due to the leading digit.

If the original behavior causes issues with a new Git cache feature that this needs to be fixed in the logic of the latter.

**Comment by s-ludwig** [comment]
One way to go about this would be to exclude cached git repositories that are not checked out on an exact tag during lookup. That probably makes sense regardless, because unexpectedly getting `1.0.0-commit.1.xxxxxxx` instead of `1.0.0-beta.1` might be just as problematic.

**Pull request merged**